include/zephyr/drivers/video.h

```diff
@@ -753,104 +753,184 @@ struct video_buffer *video_buffer_alloc(size_t size);
-
 /**
  * @defgroup video_pixel_formats Video pixel formats
+ * The @c | character separate the pixel, and spaces separate the bytes.
+ * The uppercase letter represents the most significant bit.
+ * The lowercase letters represent the rest of the bits.
  * @{
  */
 
 /**
- * @name Bayer formats
+ * @brief Four-character-code uniquely identifying the pixel format
+ */
+#define VIDEO_FOURCC(a, b, c, d)                                                                   \
+	((uint32_t)(a) | ((uint32_t)(b) << 8) | ((uint32_t)(c) << 16) | ((uint32_t)(d) << 24))
+
+/**
+ * @brief Convert a four-character string to a four-character-code
+ *
+ * This converts a string literal or variable into a four-character-code
+ * as defined by @ref VIDEO_FOURCC.
+ *
+ * @param str String to be converted
+ * @return Four-character-code.
+ */
+#define VIDEO_FOURCC_FROM_STR(str) VIDEO_FOURCC((str)[0], (str)[1], (str)[2], (str)[3])
+
+/**
+ * @name Bayer formats (R, G, B channels).
  * @{
  */
 
-/** BGGR8 pixel format */
-#define VIDEO_PIX_FMT_BGGR8 video_fourcc('B', 'G', 'G', 'R') /*  8  BGBG.. GRGR.. */
-/** GBRG8 pixel format */
-#define VIDEO_PIX_FMT_GBRG8 video_fourcc('G', 'B', 'R', 'G') /*  8  GBGB.. RGRG.. */
-/** GRBG8 pixel format */
-#define VIDEO_PIX_FMT_GRBG8 video_fourcc('G', 'R', 'B', 'G') /*  8  GRGR.. BGBG.. */
-/** RGGB8 pixel format */
-#define VIDEO_PIX_FMT_RGGB8 video_fourcc('R', 'G', 'G', 'B') /*  8  RGRG.. GBGB.. */
+/**
+ * @verbatim
+ * | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | ...
+ * | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_BGGR8      VIDEO_FOURCC('B', 'A', '8', '1')
+#define VIDEO_PIX_FMT_BGGR8_BITS 8
+
+/**
+ * @verbatim
+ * | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | ...
+ * | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_GBRG8      VIDEO_FOURCC('G', 'B', 'R', 'G')
+#define VIDEO_PIX_FMT_GBRG8_BITS 8
+
+/**
+ * @verbatim
+ * | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | ...
+ * | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_GRBG8      VIDEO_FOURCC('G', 'R', 'B', 'G')
+#define VIDEO_PIX_FMT_GRBG8_BITS 8
+
+/**
+ * @verbatim
+ * | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | ...
+ * | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_RGGB8      VIDEO_FOURCC('R', 'G', 'G', 'B')
+#define VIDEO_PIX_FMT_RGGB8_BITS 8
 
 /**
  * @}
  */
 
 /**
  * @name RGB formats
+ * Per-color (R, G, B) channels.
  * @{
  */
 
-/** RGB565 pixel format */
-#define VIDEO_PIX_FMT_RGB565 video_fourcc('R', 'G', 'B', 'P') /* 16  RGB-5-6-5 */
+/**
+ * 5-bit blue followed by 6-bit green followed by 5-bit red, in little-endian over two bytes.
+ * @verbatim
+ * | gggRrrrr | BbbbbGgg | ...
```

Comment:
I documented the BGR565-LE instead of RGB565-LE, but let's resolve the RGB565 discussion first...
https://github.com/zephyrproject-rtos/zephyr/pull/79996

Reply:
Nevermind, this is the RGB565, even if R is on the **right** in the MSB first representation, it is at `BIT(0)`, and seems to be what is expected.

Reply:
I think this is BGR565-LE.

Reply:
Endianess is about the byte address order (a chunk of 8 bits), not about the bit order (MSB or LSB first) and neither a chunk of 4 bits.

To get it simple, I think we can begin with BE first. An RGB565 in BE is represented as two bytes with the most significant **byte** is stored at the first **byte**

- RRRRRGGG(1st byte)  GGGBBBBB(2nd byte)

At this point, I think we will all agree with this. Then, LiE is basically reversing the **byte** order for a **multi-byte value**, lower significant bytes get lower address. So it becomes:

- GGGBBBBB(1st byte)  RRRRRGGG(2nd byte)

Reply:
What I documented here follows Libav and the LWN article, from where I had the representation:

If representing the format in MSBit first, B is on the left.
```
big-endian, MSBit first:
15 14 13 12 11 10 9  8  7  6  5  4  3  2  1  0
B  b  b  b  b  G  g  g  g  g  g  R  r  r  r  r 

little-endian, MSBit first:
7  6  5  4  3  2  1  0  15 14 13 12 11 10 9  8
g  g  g  R  r  r  r  r  B  b  b  b  b  G  g  g
```

![scrot_20241026_115436_453x25](https://github.com/user-attachments/assets/2a74bbcd-4087-4b6b-bbd2-a56399e33697)
![scrot_20241026_115322_450x30](https://github.com/user-attachments/assets/0ccfa15a-fff2-416a-baa1-dc129b2f1d12)


If representing the format in LSBit first, R is on the left. This representation is confusing when mixing bit-endianness and byte endianness.
```
big-endian, LSBit first:
0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15
r  r  r  r  R  g  g  g  g  g  G  b  b  b  b  B

little-endian, LSBit first:
8  9 10 11 12 13 14 15 0  1  2  3  4  5  6  7
g  g  G  b  b  b  b  B  r  r  r  r  R  g  g  g
```

> At this point, I think we all agree with this.

I am trying to understand what the major implementations did agree... and they *seems* to differ? I am really not sure as they are around since a lot of time. ^_^' Can we settle that the hardware is right to break the tie?

Libavformat *tests source* and this LWN article represent RGB565 with red on the least significant bit:
- https://lwn.net/Articles/218798/
- https://github.com/libav/libav/blob/c4642788e83b0858bca449f9b6e71ddb015dfa5d/libavfilter/vsrc_testsrc.c#L403

Gstreamer, Intel libvpl, and Libavformat *codecs* represent RGB565 with red on the most significant bit:
- https://github.com/GStreamer/gstreamer/blob/fe61bc3cee1bc7b99b550005f77d8ece3649b4ec/subprojects/gst-plugins-base/gst-libs/gst/video/video-orc.orc#L862-L864
- https://github.com/GStreamer/gstreamer/blob/fe61bc3cee1bc7b99b550005f77d8ece3649b4ec/subprojects/gst-plugins-good/ext/libcaca/gstcacatv.c#L158-L163
- https://github.com/libav/libav/blob/c4642788e83b0858bca449f9b6e71ddb015dfa5d/libavcodec/bmp.c#L163
- https://github.com/libav/libav/blob/c4642788e83b0858bca449f9b6e71ddb015dfa5d/libavcodec/xwddec.c#L180
- https://github.com/libav/libav/blob/c4642788e83b0858bca449f9b6e71ddb015dfa5d/libavcodec/dds.c#L369
- The hardware seems to follow what you say.

I think red on MSBit wins! Glad I raised this point.
A test pattern generator looks "ok" (stripes of color) even with the wrong color order, and the LWN article is just text, not code that was checked against hardware. Some deeply anchored confusion in tech it seems!
Good to have this ambiguity tamed once for all, thank you for your patience.

Committing the fix.

Reply:
Yes, I believe strongly that the LWN article is WRONG. Personally, If I need a reference, I am usually based on the kernel document. The kernel document of the latest Linux release describes the same layout about the RGB565-LE as I pointed in above comment:

https://www.kernel.org/doc/html/v6.11/userspace-api/media/v4l/pixfmt-rgb.html